File: rpc/2018yreferencias/2019/local2019PracticeData/solutions/spell/spell.java
import java.io.File;
import java.util.Scanner;

public class spell {
	public static void main(String[] args) throws Exception {
        Scanner in = new Scanner(System.in);

        int dwc = in.nextInt();
        String[] dict = new String[dwc];
        for (int k = 0; k < dwc; k++) {
            dict[k] = in.next();
        }

        int iwc = in.nextInt();
        for (int k = 0; k < iwc; k++) {
            String word = in.next();
            System.out.println(word);
            check(dict, word);            
            System.out.println();
        }
    }
    
    private static void check(String[] dict, String word) {
        for (int i = 0; i < dict.length; i++) {
            if (dict[i].equals(word)) {
                System.out.println("CORRECT");
                return;
            }
        }

        boolean foundMapping = false;
        for (String dictWord : dict) {
            if (
                oneRemoved(dictWord, word) ||
                oneAdded(dictWord, word) ||
                oneWrong(dictWord, word) ||
                transposed(dictWord, word)                
            ) {
                foundMapping = true;
            }
        }

        if (!foundMapping) {
            System.out.println("UNKNOWN");
        }
    }

    private static boolean oneRemoved(String orig, String test) {
        if (hasOmitted(orig, test)) {
            System.out.println("ONE LETTER OMITTED FROM " + orig);
            return true;
        }
        return false;
    }

    private static boolean oneAdded(String orig, String test) {
        if (hasOmitted(test, orig)) {
            System.out.println("ONE LETTER ADDED TO " + orig);
            return true;
        }
        return false;
    }

    // true if 'test' can be made by omitting any one char from
    // 'orig'
    private static boolean hasOmitted(String orig, String test) {
        if (orig.length() != test.length() + 1) {
            return false;        
        }

        int oi = 0;
        int ti = 0;

        for (oi = 0; oi < orig.length(); oi++) {
            if (ti < test.length() && orig.charAt(oi) == test.charAt(ti)) {
                ti++;
            }
        }

        return ti == oi - 1;
    }

    // true if 'test' can be made from 'orig' by changing any one
    // char.
    private static boolean oneWrong(String orig, String test) {
        if (orig.length() != test.length()) {
            return false;
        }

        int wrong = 0;
        for (int i = 0; i < orig.length(); i++) {
            if (orig.charAt(i) != test.charAt(i)) {
                wrong++;
            }
        }

        if (wrong == 1) {
            System.out.println("ONE LETTER DIFFERENT FROM " + orig);
            return true;
        }
        return false;
    }

    // true if 'test' can be made by transposing any two chars in
    // 'orig'
    private static boolean transposed(String orig, String test) {
        if (orig.length() != test.length()) {
            return false;
        }

        int wrong = 0;

        for (int i = 0; i < orig.length() - 1; i++) {
            if (orig.charAt(i) == test.charAt(i)) {
                continue;
            }

            if (
                orig.charAt(i) == test.charAt(i+1) &&
                test.charAt(i) == orig.charAt(i+1)
            ) {
                i++;
                wrong++;
                continue;
            }

            return false;
        }

        if (wrong == 1) {
            System.out.println("TWO LETTERS TRANSPOSED IN " + orig);
            return true;
        }

        return false;
    }
}

File: rpc/2018yreferencias/2019/local2019PracticeData/solutions/spell/spell.py
#!/usr/bin/env python

import sys

def one_deleted(word, dictword):
    if missing_one(dictword, word):
        print("ONE LETTER OMITTED FROM " + dictword)
        return True
    return False

def one_added(word, dictword):
    if missing_one(word, dictword):
        print("ONE LETTER ADDED TO " + dictword)
        return True
    return False

def one_changed(word, dictword):
    if len(word) != len(dictword):
        return False

    wrong = 0
    for i in range(0, len(word)):
        if word[i] != dictword[i]:
            wrong += 1

    if wrong == 1:
        print("ONE LETTER DIFFERENT FROM " + dictword)
        return True

    return False

def transposed(word, dictword):
    if len(word) != len(dictword):
        return False

    wrong = 0
    i = 0

    while i < len(dictword) - 1:
        if word[i] == dictword[i]:
            i += 1
            continue

        if word[i] == dictword[i+1] and dictword[i] == word[i+1]:
            i += 2
            wrong += 1
            continue

        return False

    if wrong == 1:
        print("TWO LETTERS TRANSPOSED IN " + dictword)
        return True

    return False

def missing_one(longer, shorter):
    if len(longer) != 1 + len(shorter):
        return False

    i = 0
    for ch in longer:
        if i < len(shorter) and shorter[i] == ch:
            i += 1

    return i + 1 == len(longer)

d = int(sys.stdin.readline())
dictwords = [sys.stdin.readline().strip() for x in range(0, d)]
n = int(sys.stdin.readline())
words = [sys.stdin.readline().strip() for x in range(0, n)]

for word in words:
    print(word)
    if word in dictwords:
        print("CORRECT")
    else:
        unknown = True
        for dw in dictwords:
            if one_deleted(word, dw) or \
                one_added(word, dw) or \
                one_changed(word, dw) or \
                transposed(word, dw):
                    unknown = False

        if unknown:
            print("UNKNOWN")
    print()

